File: OpenHarmony/developtools/hapsigner/hapsigntool_cpp/utils/src/hash_utils.cpp
/*
 * Copyright (c) 2024-2024 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <fstream>
#include <map>
#include <cstdio>
#include <cstdlib>

#include "fs_digest_utils.h"
#include "file_utils.h"
#include "hash_utils.h"

namespace OHOS {
namespace SignatureTools {
    
int HashUtils::GetHashAlgsId(const std::string& algMethod)
{
    int result = static_cast<int>(HashAlgs::USE_NONE);
    if (0 == algMethod.compare("SHA-256")) {
        result = static_cast<int>(HashAlgs::USE_SHA256);
    }
    if (0 == algMethod.compare("SHA-384")) {
        result = static_cast<int>(HashAlgs::USE_SHA384);
    }
    if (0 == algMethod.compare("SHA-512")) {
        result = static_cast<int>(HashAlgs::USE_SHA512);
    }
    return result;
}

std::string HashUtils::GetHashAlgName(int algId)
{
    if (static_cast<int>(HashAlgs::USE_SHA256) == algId) {
        return "SHA-256";
    }
    if (static_cast<int>(HashAlgs::USE_SHA384) == algId) {
        return "SHA-384";
    }
    if (static_cast<int>(HashAlgs::USE_SHA512) == algId) {
        return "SHA-512";
    }
    return "";
}

std::vector<int8_t> HashUtils::GetFileDigest(const std::string& inputFile, const std::string& algName)
{
    std::vector<int8_t> result;

    std::ifstream input(inputFile, std::ios::binary);
    if (0 != input.rdstate()) {
        PrintErrorNumberMsg("VERIFY_ERROR", VERIFY_ERROR, "failed to get input stream object!");
        return std::vector<int8_t>();
    }

    char buffer[HASH_LEN] = { 0 };
    int num = 0;
    std::map<int, std::vector<int8_t>> hashMap;

    while (!input.eof()) {
        input.read(buffer, HASH_LEN);

        if (input.fail() && !input.eof()) {
            PrintErrorNumberMsg("VERIFY_ERROR", VERIFY_ERROR, "error occurred while reading data.");
            return std::vector<int8_t>();
        }

        std::streamsize readLen = input.gcount();
        std::string str;
        for (int i = 0; i < readLen; ++i) {
            str.push_back(buffer[i]);
        }

        std::vector<int8_t> dig = GetByteDigest(str, readLen, algName);
        hashMap.emplace(num, dig);
        ++num;
    }

    if (hashMap.empty()) {
        PrintErrorNumberMsg("VERIFY_ERROR", VERIFY_ERROR, "hashMap is empty.");
        return std::vector<int8_t>();
    }

    DigestUtils fileDigestUtils(HASH_SHA256);
    for (const auto& hashMapItem : hashMap) {
        std::string str(hashMapItem.second.begin(), hashMapItem.second.end());
        fileDigestUtils.AddData(str);
    }
    std::string digest = fileDigestUtils.Result(DigestUtils::Type::BINARY);
    for (std::string::size_type i = 0; i < digest.size(); i++) {
        result.push_back(digest[i]);
    }
    return result;
}

std::vector<int8_t> HashUtils::GetDigestFromBytes(const std::vector<int8_t>& fileBytes, int64_t length,
                                                  const std::string& algName)
{
    if (fileBytes.empty() || length <= 0) {
        PrintErrorNumberMsg("VERIFY_ERROR", VERIFY_ERROR, "file bytes is empty.");
        return std::vector<int8_t>();
    }
    std::map<int, std::vector<int8_t>> hashMap;
    int64_t readLength = 0;
    int64_t num = 0;
    while (readLength < length) {
        int64_t blockLength = length - readLength > HASH_LEN ? HASH_LEN : (length - readLength);
        std::string readStr(fileBytes.begin() + readLength, fileBytes.begin() + readLength + blockLength);
        std::vector<int8_t> dig = GetByteDigest(readStr, readStr.size(), algName);
        hashMap.emplace(num, dig);
        ++num;
        readLength += readStr.size();
    }
    if (hashMap.empty()) {
        PrintErrorNumberMsg("VERIFY_ERROR", VERIFY_ERROR, "hashMap is empty.");
        return std::vector<int8_t>();
    }
    DigestUtils digestUtils(HASH_SHA256);
    for (const auto& item : hashMap) {
        std::string str(item.second.begin(), item.second.end());
        digestUtils.AddData(str);
    }
    std::string digest = digestUtils.Result(DigestUtils::Type::BINARY);
    std::vector<int8_t> result;
    for (std::string::size_type i = 0; i < digest.size(); i++) {
        result.push_back(digest[i]);
    }
    return result;
}

std::vector<int8_t> HashUtils::GetByteDigest(const std::string& str, int count, const std::string& algMethod)
{
    std::vector<int8_t> result;
    DigestUtils digestUtils(HASH_SHA256);
    digestUtils.AddData(str);
    std::string digest = digestUtils.Result(DigestUtils::Type::BINARY);
    for (std::string::size_type i = 0; i < digest.size(); i++) {
        result.push_back(digest[i]);
    }
    return result;
}

} // namespace SignatureTools
} // namespace OHOS

File: OpenHarmony/developtools/hapsigner/hapsigntool/hap_sign_tool_lib/src/main/java/com/ohos/hapsigntool/utils/HashUtils.java
/*
 * Copyright (c) 2021-2022 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.ohos.hapsigntool.utils;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStream;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.HashMap;
import java.util.Map;

/**
 * The utils function used to get hash value.
 *
 * @since 2021/12/21
 */
public class HashUtils {
    private static final LogUtils LOGGER = new LogUtils(HashUtils.class);
    private static final int HASH_LEN = 4096;
    private static final Map<String, Integer> ALG_METHOD = new HashMap<>();

    static {
        ALG_METHOD.put("SHA-224", HashAlgs.USE_SHA224);
        ALG_METHOD.put("SHA-256", HashAlgs.USE_SHA256);
        ALG_METHOD.put("SHA-384", HashAlgs.USE_SHA384);
        ALG_METHOD.put("SHA-512", HashAlgs.USE_SHA512);
    }

    /**
     * Constructor of Method
     */
    private HashUtils() {
    }

    /**
     * Get algorithm id of algorithm name.
     *
     * @param algMethod algorithm name
     * @return algorithm ID
     */
    public static int getHashAlgsId(String algMethod) {
        if (ALG_METHOD.containsKey(algMethod)) {
            return ALG_METHOD.get(algMethod);
        }
        return HashAlgs.USE_NONE;
    }

    private static MessageDigest getMessageDigest(String algMethod) {
        MessageDigest md = null;
        try {
            md = MessageDigest.getInstance(algMethod);
        } catch (NoSuchAlgorithmException e) {
            LOGGER.error("no such algorithm", e);
        }
        return md;
    }

    private static byte[] getByteDigest(byte[] str, int count, String algMethod) {
        MessageDigest md = getMessageDigest(algMethod);
        md.update(str, 0, count);
        return md.digest();
    }

    /**
     * get digest used data in file.
     *
     * @param filePath input file path.
     * @param algName algorithm name.
     * @return byte array of digest.
     */
    public static byte[] getFileDigest(String filePath, String algName) {
        byte[] digest = null;
        MessageDigest md = getMessageDigest(algName);
        try (InputStream input = new FileInputStream(new File(filePath))) {
            if (md == null) {
                throw new NoSuchAlgorithmException();
            }
            byte[] fileDate = new byte[HASH_LEN];
            int num = 0;
            int byteCount;
            HashMap<Integer, byte[]> hashList = new HashMap<Integer, byte[]>();
            while ((byteCount = input.read(fileDate)) > 0) {
                byte[] dig = getByteDigest(fileDate, byteCount, algName);
                hashList.put(num, dig);
                num++;
            }
            if (hashList.isEmpty()) {
                LOGGER.error("hashList is empty");
                return digest;
            }
            for (int i = 0; i < hashList.size(); i++) {
                md.update(hashList.get(i));
            }
            return md.digest();
        } catch (FileNotFoundException e) {
            LOGGER.error("getFileDigest File Not Found failed.");
        } catch (IOException e) {
            LOGGER.error("getFileDigest IOException failed.", e);
        } catch (NoSuchAlgorithmException e) {
            LOGGER.error("MessageDigest is null", e);
        }
        return digest;
    }

    /**
     * Algorithm of Hash.
     *
     * @since 2021/12/21
     */
    static class HashAlgs {
        /**
         * None
         */
        public static final int USE_NONE = 0;

        /**
         * The MD2 message digest.
         */
        public static final int USE_MD2 = 1;

        /**
         * The MD4 message digest.
         */
        public static final int USE_MD4 = 2;

        /**
         * The MD5 message digest.
         */
        public static final int USE_MD5 = 3;

        /**
         * The SSH-1 message digest.
         */
        public static final int USE_SHA1 = 4;

        /**
         * The SSH-224 message digest.
         */
        public static final int USE_SHA224 = 5;

        /**
         * The SSH-256 message digest.
         */
        public static final int USE_SHA256 = 6;

        /**
         * The SSH-384 message digest.
         */
        public static final int USE_SHA384 = 7;

        /**
         * The SSH-512 message digest.
         */
        public static final int USE_SHA512 = 8;

        /**
         * The RIPEMD-160 message digest.
         */
        public static final int USE_RIPEMD160 = 9;

        /**
         * Constructor of Method
         */
        private HashAlgs() {
        }

    }
}

